pkg/chunk/gcp/storage_client.go

```diff
@@ -0,0 +1,227 @@
+	cfg       Config
+	schemaCfg chunk.SchemaConfig
+	client    *bigtable.Client
+}
+
+// NewStorageClient returns a new StorageClient.
+func NewStorageClient(ctx context.Context, cfg Config, schemaCfg chunk.SchemaConfig) (chunk.StorageClient, error) {
+	client, err := bigtable.NewClient(ctx, cfg.project, cfg.instance, instrumentation())
+	if err != nil {
+		return nil, err
+	}
+	return &storageClient{
+		cfg:       cfg,
+		schemaCfg: schemaCfg,
+		client:    client,
+	}, nil
+}
+
+func (s *storageClient) NewWriteBatch() chunk.WriteBatch {
+	return bigtableWriteBatch{
+		tables: map[string]map[string]*bigtable.Mutation{},
+	}
+}
+
+type bigtableWriteBatch struct {
+	tables map[string]map[string]*bigtable.Mutation
+}
+
+func (b bigtableWriteBatch) Add(tableName, hashValue string, rangeValue []byte, value []byte) {
+	rows, ok := b.tables[tableName]
+	if !ok {
+		rows = map[string]*bigtable.Mutation{}
+		b.tables[tableName] = rows
+	}
+
+	// TODO the hashValue should actually be hashed.
+	rowKey := hashValue + separator + string(rangeValue)
+	mutation, ok := rows[rowKey]
+	if !ok {
+		mutation = bigtable.NewMutation()
+		rows[rowKey] = mutation
+	}
+
+	mutation.Set(columnFamily, column, 0, value)
+}
+
+func (s *storageClient) BatchWrite(ctx context.Context, batch chunk.WriteBatch) error {
+	bigtableBatch := batch.(bigtableWriteBatch)
+
+	for tableName, rows := range bigtableBatch.tables {
+		table := s.client.Open(tableName)
+		rowKeys := make([]string, 0, len(rows))
+		muts := make([]*bigtable.Mutation, 0, len(rows))
+		for rowKey, mut := range rows {
+			rowKeys = append(rowKeys, rowKey)
+			muts = append(muts, mut)
+		}
+
+		errs, err := table.ApplyBulk(ctx, rowKeys, muts)
+		if err != nil {
+			return err
+		}
+		for _, err := range errs {
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func (s *storageClient) QueryPages(ctx context.Context, query chunk.IndexQuery, callback func(result chunk.ReadBatch, lastPage bool) (shouldContinue bool)) error {
+	table := s.client.Open(query.TableName)
+
+	var rowRange bigtable.RowRange
+	if query.RangeValuePrefix != nil {
+		rowRange = bigtable.PrefixRange(query.HashValue + separator + string(query.RangeValuePrefix))
+	} else if query.RangeValueStart != nil {
+		rowRange = bigtable.InfiniteRange(query.HashValue + separator + string(query.RangeValueStart))
+	} else {
+		rowRange = bigtable.PrefixRange(query.HashValue + separator)
+	}
+
+	return table.ReadRows(ctx, rowRange, func(r bigtable.Row) bool {
+		// If RangeValueStart is set, BigTable doesn't know when to stop, as we're
+		// reading "until the end of the row" in DynamoDB.  So we need to check
+		// the prefix of the row is still correct.
+		if query.RangeValueStart != nil {
+			if !strings.HasPrefix(r.Key(), query.HashValue+separator) {
+				return false
+			}
+		}
+
+		return callback(bigtableReadBatch(r), false)
+	}, bigtable.RowFilter(bigtable.FamilyFilter(columnFamily)))
+}
+
+type bigtableReadBatch bigtable.Row
```

Comment:
Done - its just a side effect of smashing these two interfaces together.  BigTable returns rows one-by-one.